### PR TITLE
AEA-821 add 'build_entrypoint' field in all package configurations

### LINKED ISSUE #1710: `aea build` step in cli
**Is your feature request related to a problem? Please describe.**
Several packages contain go code which requires a build step.

**Describe the solution you'd like**
Allow package configs to specify a build step requirement and entry point. On call of `aea build` the packages with these requirements are identified and the build performed.

This prevents a situation where building takes place during launch of the agent and we can also avoid rebuilding unnecessarily.

### PR DESCRIPTION
**Pull request by marcofavorito**
## Proposed changes

Add `entrypoint:` field in package configurations.
Add `aea build` command.

## Fixes

#1710 #2021 

## Types of changes

What types of changes does your code introduce to agents-aea?
_Put an `x` in the boxes that apply_

- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist

_Put an `x` in the boxes that apply._

- [ ] I have read the [CONTRIBUTING](../CONTRIBUTING.md) doc
- [ ] I am making a pull request against the `develop` branch (left side). Also you should start your branch off our `develop`.
- [ ] Lint and unit tests pass locally with my changes
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have checked that code coverage does not decrease.
- [ ] I have checked that the documentation about the `aea cli` tool works
- [ ] I have added necessary documentation (if appropriate)
- [ ] Any dependent changes have been merged and published in downstream modules

## Further comments

n/a

### CODE REVIEW

File: aea/configurations/base.py [hunk 1]
```diff
@@ -994,6 +994,7 @@ def __init__(
         aea_version: str = "",
         fingerprint: Optional[Dict[str, str]] = None,
         fingerprint_ignore_patterns: Optional[Sequence[str]] = None,
+        entrypoint: Optional[str] = None,
```
Comment (DavidMinarsch) [inline aea/configurations/base.py:997]: Probably should be more explicitly labelled `build_entrypoint` as there might be other entry points one day..

File: aea/crypto/registries/base.py [hunk 2]
```diff
@@ -73,7 +73,7 @@ class EntryPoint(Generic[ItemType], RegexConstrainedString):
     )
 
     def __init__(self, seq):
-        """Initialize the entrypoint."""
+        """Initialize the build_entrypoint."""
```
Comment (DavidMinarsch) [inline aea/crypto/registries/base.py:76]: I don't think that's right

File: aea/cli/utils/loggers.py [hunk 3]
```diff
@@ -78,7 +78,7 @@ def _set_level(ctx, param, value):  # pylint: disable=unused-argument
             level = logging.getLevelName(value)
             logger.setLevel(level)
             # save verbosity option so it can be
-            # read in the main CLI entrypoint
+            # read in the main CLI build_entrypoint
```
Comment (DavidMinarsch) [inline aea/cli/utils/loggers.py:81]: please revert

File: aea/aea_builder.py [hunk 4]
```diff
@@ -1273,6 +1323,34 @@ def _try_to_load_agent_configuration_file(aea_project_path: Path) -> None:
...
+            "Package has not a build entrypoint specified.",
+        )
+        build_entrypoint = cast(str, build_entrypoint)
+        script_path = Path(directory) / build_entrypoint
+        enforce(
+            script_path.exists(), f"File '{build_entrypoint}' does not exists.",
+        )
+        enforce(
+            script_path.is_file(), f"'{build_entrypoint}' is not a file.",
+        )
```
Comment (DavidMinarsch) [inline aea/aea_builder.py:1346]: These are user defined values, so enforce seems not appropriate. We should probably raise an AEAException or similar.

File: aea/aea_builder.py [hunk 5]
```diff
@@ -1475,6 +1508,7 @@ def from_aea_project(
         # load agent configuration file
         configuration_file = cls.get_configuration_file_path(aea_project_path)
         agent_configuration = cls.loader.load(configuration_file.open())
+        agent_configuration.directory = configuration_file
```
Comment (DavidMinarsch) [inline aea/aea_builder.py:1511]: ?

**Review by DavidMinarsch — COMMENTED** [review]
This looks mostly very good. It would be nice if it was already implemented for the `p2p_libp2p` package as this is the first important use case of this command.

File: aea/aea_builder.py (hunk 5 above)
Comment (marcofavorito) [reply aea/aea_builder.py:1511]: not needed anymore

File: aea/aea_builder.py (hunk 4 above)
Comment (marcofavorito) [reply aea/aea_builder.py:1346]: in fact, if the condition is not satisfied, `AEAException` is raised with the error message, am I right? How would you improve this code?

File: packages/fetchai/connections/p2p_libp2p/check_dependencies.py [hunk 6]
```diff
@@ -0,0 +1,37 @@
+#!/usr/bin/env python3
```
Comment (marcofavorito) [inline packages/fetchai/connections/p2p_libp2p/check_dependencies.py:1]: relatively simple; we can iterate over it to add finer-grained checks, e.g. check the version is at least greater than some version etc.

**Review by solarw — APPROVED** [review]
LGTM

**Review by DavidMinarsch — APPROVED** [review]
LGTM

**Pull request merged**